esx_service/utils/auth_data.py

```diff
@@ -0,0 +1,568 @@
+             -- id in tenants table
+             tenant_id TEXT NOT NULL,
+             FOREIGN KEY(tenant_id) REFERENCES tenants(id)
+             );
+             '''
+        )
+
+
+        self.conn.execute(
+             '''
+             CREATE TABLE privileges(
+             -- id in tenants table
+             tenant_id TEXT NOT NULL,
+             -- datastore name
+             datastore TEXT NOT NULL,
+             -- not use currently, will drop this field later
+             global_visibility INTEGER,
+             create_volume INTEGER,
+             delete_volume INTEGER,
+             mount_volume INTEGER,
+             -- The unit of "max_volume_size" is "MB"
+             max_volume_size INTEGER,
+             -- The unit of usage_quota is "MB"
+             usage_quota INTEGER,
+             PRIMARY KEY (tenant_id, datastore),
+             FOREIGN KEY(tenant_id) REFERENCES tenants(id)
+             );
+             '''
+        )
+
+        self.conn.execute(
+             '''
+             CREATE TABLE volumes (
+             -- id in tenants table
+             tenant_id TEXT NOT NULL,
+             -- datastore name
+             datastore TEXT NOT NULL,
+             volume_name TEXT,
+             -- The unit of "volume_size" is "MB"
+             volume_size INTEGER,
+             PRIMARY KEY(tenant_id, datastore, volume_name),
+             FOREIGN KEY(tenant_id) REFERENCES tenants(id)
+             );
+             '''
+        )
+
+        self.conn.commit()
+        
+
+        
+        
+
+    def create_tenant(self, name, description, default_datastore, default_privileges, 
+                      vms, privileges):
+        """ Create a tenant in the database. 
+        
+        A tenant id will be auto-generated and returned. 
+        vms are (vm_id, vm_name) pairs. Privileges are dictionaries
+        with keys matching the row names in the privileges table. Tenant id is
+        filled in for both the vm and privileges tables.
+
+        """
+
+        # Create the entry in the tenants table
+        tenant = DockerVolumeTenant(name, description, default_datastore,
+                                    default_privileges, vms, privileges)
+        id = tenant.id
+
+        try:
+            self.conn.execute(
+                "INSERT INTO tenants(id, name, description, default_datastore) VALUES (?, ?, ?, ?)",
+                (id, name, description, default_datastore)
+            )
+
+            # Create the entries in the vms table
+            vms = [(vm_id, vm_name, id) for (vm_id, vm_name) in vms]
+                
+            if vms:
+                self.conn.executemany(
+                "INSERT INTO vms(vm_id, vm_name, tenant_id) VALUES (?, ?, ?)",
+                vms
+                )
+
+            # Create the entries in the privileges table
+            # TODO: data scrubbing, ensure all keys exist
+            for p in default_privileges:
+                p[auth_data_const.TENANT_ID] = id
```

Comment:
`default_privileges` should just be a dict and not a list. It was confusing to understand what was going on here without referring to the `get_default_datastore_and_privileges`